src/System.Private.CoreLib/src/System/HighPerformanceCounter.Windows.cs

```diff
@@ -0,0 +1,33 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System
+{
+    internal static class HighPerformanceCounter
+    {
+        public static ulong TickCount
+        {
+            get
+            {
+                Interop.Kernel32.QueryPerformanceCounter(out ulong counter);
+                return counter;
+            }
+        }
+
+        // Cache the frequency on the managed side to avoid the cost of P/Invoke on every access to Frequency
+        private static ulong s_frequency;
+
+        public static ulong Frequency
+        {
+            get
+            {
+                if (s_frequency == 0)
+                {
+                    Interop.Kernel32.QueryPerformanceFrequency(out s_frequency);
```

Comment:
This is not thread-safe for multiple reasons:
- Assignment of 64-bit numbers is not atomic on 32-bit platforms
- You do not know whether `QueryPerformanceFrequency` assigns the result just once.

Instead, you should assign it via before field init constructor. Something like:

```
private static readonly ulong s_frequency = GetFrequency();
```

It will guarantee atomicity.